src/transformers/modeling_rag.py

```diff
@@ -0,0 +1,971 @@
+                encoder_hidden_states=outputs.encoder_hidden_states,
+                encoder_attentions=outputs.encoder_attentions,
+                doc_scores=outputs.doc_scores,
+            )
+
+        logits = outputs.logits
+
+        if do_marginalize:
+            logits = self.marginalize(logits, outputs.doc_scores)
+
+        return Seq2SeqLMOutputWithDocs(
+            loss=None,
+            logits=logits,
+            past_key_values=outputs.past_key_values,
+            decoder_hidden_states=outputs.decoder_hidden_states,
+            decoder_attentions=outputs.decoder_attentions,
+            encoder_last_hidden_state=outputs.encoder_last_hidden_state,
+            encoder_hidden_states=outputs.encoder_hidden_states,
+            encoder_attentions=outputs.encoder_attentions,
+            doc_scores=outputs.doc_scores,
+        )
+
+    @torch.no_grad()
+    def generate(
+        self,
+        input_ids: Optional[torch.LongTensor] = None,
+        context_input_ids=None,
+        context_attention_mask=None,
+        retrieved_doc_embeds=None,
+        **kwargs
+    ):
+
+        if self.retriever is not None and context_input_ids is None:
+            # TODO(patrick, quention): would be nice to make retriever framework independent => we could pass `return_tensors='pt'" here
+            question_hidden_states = self.question_encoder(input_ids)[0]
+            context_input_ids, context_attention_mask, retrieved_doc_embeds = self.retriever(
+                input_ids,
+                question_hidden_states.cpu().detach().to(torch.float32),
+                prefix=self.generator.config.prefix,
+                n_docs=self.config.n_docs,
+            )
+
+            # set to correct device
+            retrieved_doc_embeds = retrieved_doc_embeds.to(question_hidden_states)
+            context_input_ids = context_input_ids.to(input_ids)
+            context_attention_mask = context_attention_mask.to(input_ids)
+
+        encoder = self.rag.generator.get_encoder()
+        encoder_outputs = encoder(input_ids=context_input_ids, attention_mask=context_attention_mask, return_dict=True)
+
+        # compute doc_scores
+        doc_scores = torch.bmm(question_hidden_states.unsqueeze(1), retrieved_doc_embeds.transpose(1, 2)).squeeze(1)
+
+        # needed to satisfy assertion that encoder_outputs.last_hidden_state.shape[0] == batch_size in generation_utils
+        last_hidden_state = encoder_outputs.last_hidden_state
+        last_hidden_state = last_hidden_state.view(-1, self.rag.config.n_docs, *last_hidden_state.shape[1:])
+
+        encoder_outputs = BaseModelOutputWithDocs(
+            last_hidden_state=last_hidden_state,
+            hidden_states=encoder_outputs.hidden_states,
+            attention_mask=context_attention_mask,
+            doc_scores=doc_scores,
+        )
+
+        kwargs["encoder_outputs"] = encoder_outputs
+        kwargs["bos_token_id"] = self.generator.config.decoder_start_token_id
```

Comment:
Hey @patrickvonplaten, I'm running e2e eval right now and running into issues with how `generate` is called here. Since we're not setting any `input_ids`, `batch_size` defaults to 1 as per logic [here](https://github.com/huggingface/transformers/blob/master/src/transformers/generation_utils.py#L283-L286), but this causes [this assertion](https://github.com/huggingface/transformers/blob/master/src/transformers/generation_utils.py#L429-L431) to fail - I can try and fix this, do you have any recommentations? Maybe pass `batch_size` argument to `generate` which will override everything?

Reply:
Very true! The last commit included a change in generation that should fix the problem. The generation method is getting really messy. This needs a bigger refactor very soon! 

Let me know, if this does not solves the problem please :-)